dex-it/src/test/scala/com/wavesplatform/it/sync/MarketOrderSuite.scala

```diff
@@ -0,0 +1,380 @@
+      val bobUBefore        = node.assetBalance(bob.toAddress.toString, UsdId.toString).balance
+      val marketPrice       = 0.1.usd
+      val marketOrderAmount = 72.waves
+      // val ordersAmount      = 36.waves
+
+      placeOrders(alice, wavesUsdPair, OrderType.BUY)(
+        10.waves -> 0.2.usd,
+        14.waves -> 0.3.usd,
+        12.waves -> 0.4.usd,
+      )
+
+      node.waitOrderProcessed(
+        wavesUsdPair,
+        node.placeMarketOrder(node.prepareOrder(bob, wavesUsdPair, OrderType.SELL, marketOrderAmount, marketPrice, fee)).message.id)
+
+      val orderBook = node.orderBook(wavesUsdPair)
+      orderBook.bids should be(empty)
+      orderBook.asks should be(empty)
+
+      /* market order fee value depends on matched orders in proportion
+       *
+       *  TODO: bug DEX-454
+       *  node.accountBalances(bob.toAddress.toString)._1 should be(bobWBefore - ordersAmount - fee * ordersAmount / marketOrderAmount)
+       */
+
+      node.assetBalance(bob.toAddress.toString, UsdId.toString).balance should be(
+        bobUBefore + 0.2.usd * 10.waves / 1.waves + 0.3.usd * 14.waves / 1.waves + 0.4.usd * 12.waves / 1.waves)
+
+    }
+
+    "should be removed from order book when there are no suitable orders by limit of the price (BUY)" in {
+      val accountBalanceWBefore = 1.waves
+      val marketOrderAmount     = 200.waves
+      val marketPrice           = 0.1.usd
+      val anotherOrderAmount    = 1.waves
+
+      val account = createAccountWithBalance()(
+        1.waves -> None,
+        100.usd -> Some(UsdId.toString)
+      )
+
+      val creationTime = System.currentTimeMillis();
+
+      node.placeOrder(node.prepareOrder(alice, wavesUsdPair, OrderType.SELL, 200.waves, marketPrice, fee)).message.id
+      node.placeOrder(node.prepareOrder(bob, wavesUsdPair, OrderType.BUY, 1.waves, marketPrice, fee * 2, creationTime = creationTime)).message.id
+      node.waitOrderProcessed(
+        wavesUsdPair,
+        node
+          .placeMarketOrder(node.prepareOrder(account, wavesUsdPair, OrderType.BUY, marketOrderAmount, marketPrice, fee, creationTime = creationTime))
+          .message
+          .id
+      )
+
+      val orderBook = node.orderBook(wavesUsdPair)
+      orderBook.bids should be(empty)
+      orderBook.asks should be(empty)
+
+      node.accountBalances(account.toAddress.toString)._1 should be(
+        accountBalanceWBefore + marketOrderAmount - anotherOrderAmount - fee * (marketOrderAmount - anotherOrderAmount) / marketOrderAmount)
+    }
+  }
+
+  "Validation market orders" - {
+    val price = 0.4.usd
+
+    "should be accepted if there is no way to fill it completely (sum of all orders in order book < amount of market order" in {
+      placeOrders(bob, wavesUsdPair, OrderType.SELL)(100.waves -> price)
+
+      var total = 0L
+      node
+        .orderBook(wavesUsdPair)
+        .bids
+        .foreach(order => {
+          total += order.amount
+        })
```

Comment:
More functional way to solve this task may look like this:
```
val total = node.orderBook(wavesUsdPair).bids.map(_.amount).sum
```
We try to avoid using of `var`s

Reply:
tnx